### PR TITLE
CORE-6858 Uniqueness checker error handling refinement

### PR DESCRIPTION
**Pull request by relyafi**
### Summary
This PR makes improvements to the uniqueness checker error handling logic. Prior to this change, the uniqueness checker did not have a way of meaningfully communicating "exceptional" errors back to the client, i.e. errors that are not for business reasons such as double spend attempts, time window out of bounds etc.

In reality, there will always be some scenarios where the uniqueness checker needs to handle unexpected exceptions. This PR changes the behavior of the uniqueness checker so that:

- Expected failures continue to return a `UniquenessCheckResponse` with a specific failure result type
- Unexpected failures, i.e. exceptions that cannot be caught and handled gracefully by the uniqueness checker will now be mapped to a new `UniquenessCheckResultUnhandledExceptionAvro` object

In practice, the new object will not be returned to the client; the uniqueness message processor will instead convert such messages into a platform error result on the external events response, and will forward an `ExceptionEnvelope` with the error details. All other result types will continue to return success results (indicating the operation was successfully processed, as opposed to the transaction itself being successful).

### Testing

New unit and message bus integration tests have been added to verify the expected behavior. At present, there is no easy way to force an exception to be raised reliably in integration tests, so the message bus integration tests use a customised fake backing store that can be made to throw an exception on demand.

It is likely that following implementation of dynamic uniqueness DB support (in a subsequent PR), it will be possible to force exceptions to be thrown by specifying a holding identity that does not have an associated uniqueness DB, which will remove the need for the customised fake object.

### CODE REVIEW

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerImpl.kt [hunk 1]
```diff
@@ -153,24 +156,39 @@ class BatchedUniquenessCheckerImpl(
...
+                    when (result) {
+                        is UniquenessCheckResultSuccess -> {
+                            UniquenessCheckResultSuccessAvro(result.resultTimestamp)
+                        }
+                        is UniquenessCheckResultFailure -> {
+                            result.toExternalError()
+                        }
+                        else -> {
+                            throw IllegalStateException(
+                                "Unknown result type: ${result.javaClass.typeName}"
```
Comment (r3hj) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerImpl.kt:172]: Perhaps, it'd be more useful for debugging if the exception has the request information.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt [hunk 2]
```diff
@@ -23,9 +24,13 @@ class UniquenessCheckMessageProcessor(
         val requests = events.map { it.value }.filterNotNull()
```
Comment (r3hj) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: Minor suggestion though this change is not part of this PR:

```val requests = events.mapNotNull { it.value }```

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt [hunk 3]
```diff
@@ -23,9 +24,13 @@ class UniquenessCheckMessageProcessor(
...
 
         return requests.zip(uniquenessChecker.processRequests(requests)).map { (request, response) ->
-            // TODO - need to rethink exception handling and whether we should be throwing exceptions
-            // with platform errors or not
-            externalEventResponseFactory.success(request.flowExternalEventContext, response)
+            if (response.result !is UniquenessCheckResultUnhandledExceptionAvro) {
```
Comment (r3hj) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: A rather subjective view: inverting the check may improve the code readability. e.g. `if(response.result is ...`) {

Comment (nkovacsx) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: NIT: I personally prefer having the non-negated condition in the if statement, i.e. if (response.result is UniquenessCheckResultUnhandledExceptionAvro)

**Review by nkovacsx — DISMISSED** [review]
One minor comment, but approving because it's only a cosmetic stuff

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt (hunk 3 above)
Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: Yeah, I can see how that might be a little clearer (it might be easy to gloss over the not part and subconsciously read it as is). Think probably not worth changing and having to retest etc though.

**Review by mnesbit — DISMISSED** [review]
I defer to @lankydan but I am fine with this structurally

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt (hunk 3 above)
Comment (lankydan) [inline components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: Why do we need `UniquenessCheckResultUnhandledExceptionAvro` as an Avro object if we don't send it? Are we receiving it from somewhere else? Similar thought about another object somewhere else.

**Comment by r3hj** [comment]
LGTM

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt (hunk 3 above)
Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: @lankydan there's really two reasons. First, we use the "Avro" objects in the uniqueness checker API, but don't expose external events on this API. So we need something to communicate the error without immediately raising a platform error. This might not be a good enough reason in and of itself, but also, I'm looking ahead to the point where maybe we want to invoke the uniqueness checker in another way than from the flow worker using the external events API. In this scenario, we'd still be using the Avro objects I imagine (since that's what the uniqueness checker API talks in), but we may not have the message processor + external events API available.

Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/UniquenessCheckMessageProcessor.kt]: Ok, well I'm now outnumbered on this! Will address.

File: components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerImpl.kt (hunk 1 above)
Comment (relyafi) [reply components/uniqueness/uniqueness-checker-impl/src/main/kotlin/net/corda/uniqueness/checker/impl/BatchedUniquenessCheckerImpl.kt:172]: I think the full request might be overkill. We could add the txId but this might be a duplicate. I'd say this should in practice be an unreachable branch, so feel on balance it's probably not worth worrying about right now.

**Review by nkovacsx — DISMISSED** [review]
🆗

**Review by lankydan — DISMISSED** [review]

**Review by mnesbit — APPROVED** [review]
OK
